vortex-serde/src/io/file.rs

```diff
@@ -0,0 +1,130 @@
+use std::fs::File;
+use std::io;
+use std::os::fd::{FromRawFd, IntoRawFd, RawFd};
+use std::os::unix::fs::FileExt;
+use std::path::Path;
+use std::sync::Arc;
+
+use vortex_error::vortex_panic;
+
+use super::VortexReadAt;
+
+/// A wrapper on top of a [`File`] that provides shared, immutable access to
+/// the underlying file descriptor.
+#[derive(Debug, Clone)]
+pub struct SharedFile(Arc<Inner>);
+
+// Implement an inner file.
+#[derive(Debug, Clone)]
+struct Inner {
+    raw_fd: RawFd,
+}
+
+impl SharedFile {
+    /// Open a file on the current file system.
+    ///
+    /// The `SharedFile` takes ownership of the file descriptor, and
+    /// can lend out immutable handles to it for multiple operations, without
+    /// needing to open a new handle.
+    pub fn open(path: impl AsRef<Path>) -> io::Result<Self> {
+        let f = File::open(path)?;
+        let raw_fd = f.into_raw_fd();
+        let inner = Arc::new(Inner { raw_fd });
+
+        Ok(Self(inner))
+    }
+}
+
+// Provide access to a file operation directly.
+//
+// We want to implement all of the extension functions via this read_at.
+fn shared_read_exact_at(file: SharedFile, buf: &mut [u8], offset: u64) -> io::Result<()> {
+    let fd = file.0.raw_fd;
+
+    // SAFETY: fd is owned by the SharedFile, file is not closed except when SharedFile drops.
+    let std_file = unsafe { File::from_raw_fd(fd) };
+    let res = std_file.read_exact_at(buf, offset);
+
+    // consume the file descriptor so that we do not close the file
+    std_file.into_raw_fd();
+
+    res
+}
+
+// TODO: change all of the size() methods to return an io::Result<u64>
+fn shared_fd_size(file: SharedFile) -> u64 {
+    let fd = file.0.raw_fd;
+    // SAFETY: fd is owned by the SharedFile, file is not closed except when SharedFile drops.
+    let std_file = unsafe { File::from_raw_fd(fd) };
+    let res = std_file
+        .metadata()
+        .map(|m| m.len())
+        .unwrap_or_else(|e| vortex_panic!("read SharedFile size: {e}"));
```

Comment:
we should `into_raw_fd` before we unwrap, just in case this is called by something that has a panic hook installed and the panic doesn't abort the process

Reply:
but what we _actually_ should do is update this method to return io::Result<u64>